diff --git a/src/alloc.c b/src/alloc.c
index c77bdc6372d..f455260c3b0 100644
--- a/src/alloc.c
+++ b/src/alloc.c
@@ -1087,11 +1087,7 @@ lisp_free (void *block)
    BLOCK_BYTES and guarantees they are aligned on a BLOCK_ALIGN boundary.  */

 /* Byte alignment of storage blocks.  */
-#ifdef HAVE_UNEXEC
-# define BLOCK_ALIGN (1 << 10)
-#else  /* !HAVE_UNEXEC */
-# define BLOCK_ALIGN (1 << 15)
-#endif
+#define BLOCK_ALIGN (1 << 15)
 static_assert (POWER_OF_2 (BLOCK_ALIGN));

 /* Use aligned_alloc if it or a simple substitute is available.
